pkg/image/validation.go

```diff
@@ -0,0 +1,153 @@
+package image
+
+import (
+	"fmt"
+	"strings"
+)
+
+func validate(definition *Definition) error {
+	err := validateAPIVersion(definition)
+	if err != nil {
+		return fmt.Errorf("error validating apiVersion: %w", err)
+	}
+	err = validateImage(definition)
+	if err != nil {
+		return fmt.Errorf("error validating image: %w", err)
+	}
+	err = validateOperatingSystem(definition)
+	if err != nil {
+		return fmt.Errorf("error validating operating system: %w", err)
+	}
+
+	return nil
+}
+
+func validateAPIVersion(definition *Definition) error {
+	// Open to discussion regarding how to validate the else if
+	if definition.APIVersion == (Definition{}.APIVersion) {
+		return fmt.Errorf("apiVersion not defined")
+	} else if len(definition.APIVersion) < 2 {
+		return fmt.Errorf("invalid apiVersion")
+	}
+	return nil
+}
+
+func validateImage(definition *Definition) error {
+	if definition.Image == (Image{}) {
+		return fmt.Errorf("image not defined")
+	}
+	if definition.Image.ImageType == (Definition{}.Image.ImageType) {
+		return fmt.Errorf("imageType not defined")
+	} else if definition.Image.ImageType != "iso" && definition.Image.ImageType != "raw" {
+		return fmt.Errorf("invalid imageType, should be 'iso' or 'raw'")
+	}
+	if definition.Image.BaseImage == (Definition{}.Image.BaseImage) {
+		return fmt.Errorf("baseImage not defined")
+	}
+	if definition.Image.OutputImageName == (Definition{}.Image.OutputImageName) {
+		return fmt.Errorf("outputImageName not defined")
+	}
+	return nil
+}
+
+func validateOperatingSystem(definition *Definition) error {
+	if checkIfOperatingSystemDefined(&definition.OperatingSystem) {
+		return nil
+	}
+	err := validateKernalArgs(&definition.OperatingSystem)
+	if err != nil {
+		return fmt.Errorf("error validating kernal args: %w", err)
+	}
+	err = validateSystemd(&definition.OperatingSystem)
+	if err != nil {
+		return fmt.Errorf("error validating systemd args: %w", err)
+	}
+	err = validateUsers(&definition.OperatingSystem)
+	if err != nil {
+		return fmt.Errorf("error validating users: %w", err)
+	}
+	return nil
+}
+
+func checkIfOperatingSystemDefined(os *OperatingSystem) bool {
+	return len(os.KernelArgs) == 0 &&
+		len(os.Users) == 0 &&
+		len(os.Systemd.Enable) == 0 && len(os.Systemd.Disable) == 0 &&
+		os.Suma.Host == "" && os.Suma.ActivationKey == "" && !os.Suma.GetSSL
+}
+
+func validateKernalArgs(os *OperatingSystem) error {
+	seenKeys := make(map[string]bool)
+
+	for _, arg := range os.KernelArgs {
+		parts := strings.SplitN(arg, "=", 2)
```

Comment:
Could we please verify if this is always the case for argument layouts? I'm almost fairly certain that there are args which can be just "keys" but I might be wrong.

Reply:
For my next commit containing tests, I'm gonna operate under the "key=value" assumption and then just change it later if necessary.

Reply:
Can confirm, some kernel cmdline arguments are just parameters, without value. See my kernel cmdline for example, where it has "quiet" as a parameter/key, with no required value:

```
rdo@eostre:~% cat /proc/cmdline
BOOT_IMAGE=/boot/vmlinuz-6.6.7-1-default root=UUID=8cfd88e7-a472-4edf-8845-fc7a28b611d5 splash=silent resume=/dev/disk/by-uuid/06b2a538-d922-4c75-a5be-ec0660a185fd mitigations=auto quiet security=selinux selinux=1 enforcing=1
```